### PR TITLE
Prepare modules ec2_vpc_nacl and ec2_vpc_nacl_info for promotion

### PR DESCRIPTION
**Pull request by abikouo**
##### SUMMARY

- Use shared code from `amazon.aws.plugins.module_utils.ec2`
- Refactor module ec2_vpc_nacl and ec2_vpc_nacl_info

##### ISSUE TYPE

- Feature Pull Request

### CODE REVIEW

File: plugins/modules/ec2_vpc_nacl.py [hunk 1]
```diff
@@ -255,42 +253,102 @@ def tags_changed(nacl_id, client, module):
...
+            boto3_rule["CidrBlock"] = ansible_rule[3]
+        if icmp_present(ansible_rule):
+            boto3_rule["IcmpTypeCode"] = {"Type": int(ansible_rule[4]), "Code": int(ansible_rule[5])}
+        else:
+            if ansible_rule[6] or ansible_rule[7]:
+                boto3_rule["PortRange"] = {"From": ansible_rule[6], "To": ansible_rule[7]}
+    return boto3_rule
+
+
+def diff_network_acl_rules(
```
Comment (alinabuzachis) [inline plugins/modules/ec2_vpc_nacl.py:275]: What about splitting this function into one for finding all rules to add and the other for finding all rules to remove?

File: plugins/modules/ec2_vpc_nacl.py [hunk 2]
```diff
@@ -255,42 +253,102 @@ def tags_changed(nacl_id, client, module):
...
+                rule_action=rule_action,
+                rule_number=rule_number,
+                **rule,
+            )
+
+    # Removed Rules
+    for rule in removed_rules:
+        changed = True
+        if not module.check_mode:
+            delete_network_acl_entry(client, network_acl_id=nacl_id, rule_number=rule["RuleNumber"], egress=egress)
```
Comment (alinabuzachis) [inline plugins/modules/ec2_vpc_nacl.py:341]: Try-except?

File: plugins/modules/ec2_vpc_nacl.py [hunk 3]
```diff
@@ -255,42 +253,102 @@ def tags_changed(nacl_id, client, module):
...
+            if existing["RuleNumber"] == rule["RuleNumber"] and existing == rule:
+                _match = True
+                break
+        if not _match:
+            rules_to_remove.append(existing)
+
+    return rules_to_add, rules_to_remove
+
+
+def rules_changed(
```
Comment (alinabuzachis) [inline plugins/modules/ec2_vpc_nacl.py:304]: What about splitting this function into one for finding added rules and another for finding removed rules?

File: plugins/modules/ec2_vpc_nacl.py [hunk 4]
```diff
@@ -300,306 +358,208 @@ def process_rule_entry(entry, Egress):
...
-    for rule in module.params.get("egress"):
-        params = process_rule_entry(rule, Egress=True)
-        params["NetworkAclId"] = nacl["NetworkAcl"]["NetworkAclId"]
-        create_network_acl_entry(params, client, module)
+    return params
 
 
-# Module invocations
-def setup_network_acl(client, module):
+def construct_acl_entries(client, module: AnsibleAWSModule, nacl_id: str) -> bool:
```
Comment (alinabuzachis) [inline plugins/modules/ec2_vpc_nacl.py:365]: What about splitting this one too (one for egress and one for ingress)?

File: plugins/modules/ec2_vpc_nacl.py [hunk 5]
```diff
@@ -170,70 +175,63 @@
...
-    if subs_added:
-        replace_network_acl_association(nacl_id, subs_added, client, module)
-        changed = True
-    subs_removed = subnets_removed(nacl_id, subnets, client, module)
-    if subs_removed:
-        default_nacl_id = find_default_vpc_nacl(vpc_id, client, module)[0]
-        replace_network_acl_association(default_nacl_id, subs_removed, client, module)
-        changed = True
+
+    network_acls = describe_network_acls(client, NetworkAclIds=[nacl_id])
```
[the remark this answers is not in the document]
Comment (abikouo) [reply plugins/modules/ec2_vpc_nacl.py:201]: same as above

File: plugins/modules/ec2_vpc_nacl.py (hunk 2 above)
Comment (abikouo) [reply plugins/modules/ec2_vpc_nacl.py:341]: same as all exceptions

File: plugins/modules/ec2_vpc_nacl.py (hunk 1 above)
Comment (abikouo) [reply plugins/modules/ec2_vpc_nacl.py]: done

File: plugins/modules/ec2_vpc_nacl.py (hunk 3 above)
Comment (abikouo) [reply plugins/modules/ec2_vpc_nacl.py:285]: I have simplified it, but there is no added value to split between the added and the removed rules, this is done by the finding process

File: plugins/modules/ec2_vpc_nacl.py (hunk 4 above)
Comment (abikouo) [reply plugins/modules/ec2_vpc_nacl.py]: done

File: plugins/modules/ec2_vpc_nacl_info.py [hunk 6]
```diff
@@ -121,55 +122,60 @@
...
     try:
-        nacls = connection.describe_network_acls(aws_retry=True, NetworkAclIds=nacl_ids, Filters=filters)
-    except is_boto3_error_code("InvalidNetworkAclID.NotFound"):
-        module.fail_json(msg="Unable to describe ACL.  NetworkAcl does not exist")
-    except (
-        botocore.exceptions.ClientError,
-        botocore.exceptions.BotoCoreError,
-    ) as e:  # pylint: disable=duplicate-except
+        snaked_nacls = [format_nacl(nacl) for nacl in describe_network_acls(connection, **params)]
+    except is_ansible_aws_error_code("InvalidNetworkAclID.NotFound"):
```
Comment (GomathiselviS) [inline plugins/modules/ec2_vpc_nacl_info.py:170]: Shouldn't this be handled by a.aws' module_utils.ec2?

File: tests/integration/targets/ec2_vpc_nacl/tasks/ingress_and_egress.yml [hunk 7]
```diff
@@ -1,6 +1,6 @@
 # ============================================================
 - block:
-  - name: create ingress and egress rules using subnet IDs
+  - name: Create ingress and egress rules using subnet IDs
```
Comment (GomathiselviS) [inline tests/integration/targets/ec2_vpc_nacl/tasks/ingress_and_egress.yml:3]: can you use fqcn for all the module calls?

**Review by GomathiselviS — APPROVED** [review]

**Comment by abikouo** [comment]
regate

**Pull request merged**